src/core/morphing.js

```diff
@@ -0,0 +1,110 @@
+import { Idiomorph } from "idiomorph/dist/idiomorph.esm.js"
+import { FrameElement } from "../elements/frame_element"
+import { dispatch } from "../util"
+
+export function morphRefresh(currentElement, newElement) {
+  idiomorph(currentElement, newElement, {
+    shouldSkipMorphing: element => canRefreshFrame(element)
+  })
+
+  for (const frame of document.querySelectorAll("turbo-frame")) {
+    if (canRefreshFrame(frame)) refreshFrame(frame)
+  }
+
+  dispatch("turbo:morph", { detail: { currentElement, newElement } })
+}
+
+export function morphFrames(currentElement, newElement) {
+  dispatch("turbo:before-frame-morph", {
+    target: currentElement,
+    detail: { currentElement, newElement }
+  })
+
+  morphChildren(currentElement, newElement)
+}
+
+export function morphChildren(currentElement, newElement) {
+  idiomorph(currentElement, newElement.children, {
+    options: {
+      morphStyle: "innerHTML"
+    }
+  })
+}
+
+export function morphElements(currentElement, newElement) {
+  idiomorph(currentElement, newElement)
+}
+
+const defaultOptions = {
+  morphStyle: "outerHTML"
+}
+
+function idiomorph(currentElement, newElement, delegate = {}) {
```

Comment:
I found the double *delegate* nomenclature a bit confusing at first. What do you think of making our `idiomorph` to receive just `options`:

```js
idiomorph(currentElement, newElement, options = {})
```

Then, we could have a default options

```js
const defaultOptions = {
  morphStyle: "outerHTML",
  shouldSkipMorphing: (element) => false
}
```

Even when `shouldSkipMorphing` is our property but not idiomorph's; I think that's fine, as we offer a unified interface to the caller. Then, you could remove the `invoke` function since there is no conditional anymore, and you can invoke `this.delegate[shouldSkipMorphing](target)` in a deterministic way.

Reply:
I like it! I've pushed up a change that changes the terminology. I've also changed "shouldSkipMorphing" to "willMorphElement", since that feels more communicative and less circuitous.